sdk/identity/identity/src/msal/nodeFlows/msalClientCertificate.ts

```diff
@@ -46,59 +46,70 @@ interface CertificateParts {
-    super(options);
-    this.requiresConfidential = true;
-    this.sendCertificateChain = options.sendCertificateChain;
-
-    const parts = this.parseCertificate(options.certificatePath);
-    this.msalConfig.auth.clientCertificate = {
-      thumbprint: parts.thumbprint,
-      privateKey: parts.certificateContents,
-      x5c: parts.x5c
-    };
+export function parseCertificate(
+  certificatePath: string,
+  sendCertificateChain?: boolean
+): CertificateParts {
+  const certificateParts: Partial<CertificateParts> = {};
+
+  certificateParts.certificateContents = readFileSync(certificatePath, "utf8");
+  if (sendCertificateChain) {
+    certificateParts.x5c = certificateParts.certificateContents;
   }
 
-  private parseCertificate(certificatePath: string): CertificateParts {
-    const certificateParts: Partial<CertificateParts> = {};
+  const certificatePattern = /(-+BEGIN CERTIFICATE-+)(\n\r?|\r\n?)([A-Za-z0-9+/\n\r]+=*)(\n\r?|\r\n?)(-+END CERTIFICATE-+)/g;
+  const publicKeys: string[] = [];
 
-    certificateParts.certificateContents = readFileSync(certificatePath, "utf8");
-    if (this.sendCertificateChain) {
-      certificateParts.x5c = certificateParts.certificateContents;
+  // Match all possible certificates, in the order they are in the file. These will form the chain that is used for x5c
+  let match;
+  do {
+    match = certificatePattern.exec(certificateParts.certificateContents);
+    if (match) {
+      publicKeys.push(match[3]);
     }
+  } while (match);
 
-    const certificatePattern = /(-+BEGIN CERTIFICATE-+)(\n\r?|\r\n?)([A-Za-z0-9+/\n\r]+=*)(\n\r?|\r\n?)(-+END CERTIFICATE-+)/g;
-    const publicKeys: string[] = [];
+  if (publicKeys.length === 0) {
+    const error = new Error(
+      "The file at the specified path does not contain a PEM-encoded certificate."
+    );
+    throw error;
+  }
 
-    // Match all possible certificates, in the order they are in the file. These will form the chain that is used for x5c
-    let match;
-    do {
-      match = certificatePattern.exec(certificateParts.certificateContents);
-      if (match) {
-        publicKeys.push(match[3]);
-      }
-    } while (match);
+  certificateParts.thumbprint = createHash("sha1")
+    .update(Buffer.from(publicKeys[0], "base64"))
+    .digest("hex")
+    .toUpperCase();
 
-    if (publicKeys.length === 0) {
-      const error = new Error(
-        "The file at the specified path does not contain a PEM-encoded certificate."
-      );
+  return certificateParts as CertificateParts;
+}
+
+/**
+ * MSAL client certificate client. Calls to MSAL's confidential application's `acquireTokenByClientCredential` during `doGetToken`.
+ * @internal
+ */
+export class MsalClientCertificate extends MsalNode {
+  constructor(options: MSALClientCertificateOptions) {
+    super(options);
+    this.requiresConfidential = true;
+    try {
+      const parts = parseCertificate(options.certificatePath, options.sendCertificateChain);
```

Comment:
I see, we're making this function sync because it's used in a constructor... is there any way to avoid this? Can it be passed in somehow instead? Or computed lazily by getToken?

Reply:
I’ve pushed a change that makes this async, but keep in mind that’s technically a breaking change…

Which we can get away with, since we’re releasing v2 in a month 🥳

So, I think it’s ok 🤔